feat: Add custom placeholder responses for scale-from-zero scenarios

malpou opened the pull request:
Allows `HTTPScaledObjects` to serve configurable content (HTML, JSON, XML, plain text, etc.) while workloads scale up from zero, improving user experience during cold starts.

## Description

This PR implements customizable placeholder responses that are displayed to users while applications are scaling up from zero. This significantly improves the user experience during cold starts by providing immediate feedback instead of connection timeouts or errors.

The implementation is **content-agnostic**, allowing users to serve any type of response (HTML pages, JSON API responses, XML, plain text, etc.) without automatic modifications or assumptions about the content format.

## Key features:

- Adds `placeholderConfig` section to `HTTPScaledObject` CRD for configuring placeholder responses
- **Content-type agnostic**: Support for any response format (HTML, JSON, XML, plain text, etc.)
- Inline content or ConfigMap-based templates via environment variables
- Go template support with variables like `{{.ServiceName}}`, `{{.Namespace}}`, and `{{.RefreshInterval}}`
- Custom `Content-Type` headers to match your API format
- Custom HTTP status codes (default: 200 OK)
- No automatic content injection - your content is served exactly as provided
- Configurable via global defaults (ConfigMap) or per-HTTPScaledObject settings

## Implementation details:

- New placeholder handler in the interceptor that serves configured content when backends are unavailable
- Template variable substitution works with any content format
- Graceful fallback to upstream requests once the service is ready
- Error handling for endpoint cache failures returns 503 Service Unavailable
- Full backward compatibility - disabled by default

## Use cases:

- **HTML pages**: Show a "Service starting..." page with auto-refresh
- **JSON APIs**: Return `{"status": "initializing", "message": "Service is starting..."}` responses
- **XML services**: Provide XML-formatted status responses
- **Plain text**: Simple text-based status messages

## Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO)
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [x] Any necessary documentation is added, such as:
  - [`README.md`](/README.md)
  - [The `docs/` directory](./docs)
  - [The docs repo](https://github.com/kedacore/keda-docs)

malpou committed:
feat: Add custom placeholder pages for scale-from-zero scenarios

Allows HTTPScaledObjects to serve configurable HTML pages while workloads
scale up from zero, with support for templates, custom headers, and
automatic refresh.

Signed-off-by: malpou <malthe@grundtvigsvej.dk>

JorTurFer commented on interceptor/handler/placeholder.go:
```diff
@@ -0,0 +1,215 @@
+}
+
+// getTemplate retrieves the template for the given HTTPScaledObject
+func (h *PlaceholderHandler) getTemplate(ctx context.Context, hso *v1alpha1.HTTPScaledObject) (*template.Template, error) {
```
Should we include in the something like the HTTPScaledObject and/or the configMap generation somewhere? we need a mechanism to reevaluate the template if there is a change, so maybe storing the resource generations in the cache can help to enforce the cache regeneration when there are changes

JorTurFer commented on interceptor/handler/placeholder.go:
```diff
@@ -0,0 +1,215 @@
+}
+
+// ClearCache clears the template cache
+func (h *PlaceholderHandler) ClearCache() {
```
Is this really needed? I mean, when we would clear the whole cache?

malpou replied:
It's been removed since it was only used in the tests

JorTurFer commented on interceptor/proxy_handlers.go:
```diff
@@ -69,6 +72,25 @@ func newForwardingHandler(
+		if placeholderHandler != nil && httpso.Spec.PlaceholderConfig != nil && httpso.Spec.PlaceholderConfig.Enabled {
+			endpoints, err := endpointsCache.Get(httpso.GetNamespace(), httpso.Spec.ScaleTargetRef.Service)
+			if err == nil && workloadActiveEndpoints(endpoints) == 0 {
+				lggr.Info("serving placeholder page immediately",
```
Let's make this a debug trace

JorTurFer commented on tests/checks/placeholder_pages/placeholder_pages_test.go:
```diff
@@ -0,0 +1,238 @@
+      labels:
+        app: {{.TestName}}
+    spec:
+      containers:
+      - name: app
+        image: kennethreitz/httpbin:latest
+        ports:
+        - containerPort: 80
+          name: http
+        resources:
+          requests:
+            memory: "64Mi"
+            cpu: "100m"
+          limits:
+            memory: "128Mi"
+            cpu: "200m"
```
let's use the same image as for other e2e tests
```suggestion
      containers:
      - name: {{.TestName}}
        image: registry.k8s.io/e2e-test-images/agnhost:2.45
        args:
        - netexec
        ports:
        - name: http
          containerPort: 8080
          protocol: TCP
        readinessProbe:
          httpGet:
            path: /
            port: http
```

JorTurFer commented on tests/checks/placeholder_pages/placeholder_pages_test.go:
```diff
@@ -0,0 +1,238 @@
+spec:
+  ports:
+  - port: 80
+    targetPort: 80
```
```suggestion
    targetPort: 8080
```

JorTurFer commented on tests/checks/placeholder_pages/placeholder_pages_test.go:
```diff
@@ -0,0 +1,238 @@
+		DeleteNamespace(t, testNamespace)
+	}()
+
+	assert.True(t, WaitForDeploymentReplicaReadyCount(t, KubeClient, testName, testNamespace, 1, 240, 3),
+		"deployment replicas should be 1 after 3 minutes")
+
+	// Wait for the deployment to scale down to zero
+	time.Sleep(90 * time.Second)
+	assert.True(t, WaitForDeploymentReplicaReadyCount(t, KubeClient, testName, testNamespace, 0, 240, 3),
+		"deployment replicas should be 0 after 3 minutes")
```
let's deploy this with 0 instances since the beginning to save time
Do we need to deploy it with 1 instance and wait until it scales down for any specific reason?

malpou replied:
Adjusted

JorTurFer commented on tests/checks/placeholder_pages/placeholder_pages_test.go:
```diff
@@ -0,0 +1,238 @@
+	t.Log("Testing placeholder page response")
+
+	var interceptorIP string
+	if UseIngressHost {
+		interceptorIP = ExternalIP
+	} else {
+		interceptorService := GetKubernetesServiceEndpoint(
+			t,
+			KubeClient,
+			"keda-http-add-on-interceptor-proxy",
+			"keda",
+		)
+		interceptorIP = interceptorService.IP
+	}
```
I think that this is a leftover from a local test

malpou replied:
Yes, that was a leftover

JorTurFer reviewed (commented):
This is a great feature! Thanks for the contribution :bow:

malpou committed:
remove ClearCache funtion

Signed-off-by: malpou <malthe@grundtvigsvej.dk>

malpou committed:
use same image for e2e tests

Signed-off-by: malpou <malthe@grundtvigsvej.dk>

JorTurFer commented on interceptor/handler/placeholder.go:
```diff
@@ -0,0 +1,215 @@
+<html>
+<head>
+    <title>Service Starting</title>
+    <meta http-equiv="refresh" content="{{.RefreshInterval}}">
```
@wozniakjan @zroubalik WDYT?+

malpou committed:
adjust tests according to comments

Signed-off-by: malpou <malthe@grundtvigsvej.dk>

malpou committed:
add RWMutex to ensure that the template cache is locked when being written to

Signed-off-by: malpou <malthe@grundtvigsvej.dk>

malpou committed:
adjust e2e tests away from my local setup and over to something that should be comptatible in the GHA environment

Signed-off-by: malpou <malthe@grundtvigsvej.dk>

malpou committed:
Update placeholder_pages_test.go

Signed-off-by: Malthe Poulsen <30603252+malpou@users.noreply.github.com>
Signed-off-by: malpou <malthe@grundtvigsvej.dk>

malpou committed:
take on injecting reloading watching js using HEAD requests

Signed-off-by: malpou <malthe@grundtvigsvej.dk>

malpou committed:
remove outdated text

Signed-off-by: malpou <malthe@grundtvigsvej.dk>

malpou committed:
fix unit tests

Signed-off-by: Malthe Poulsen <30603252+malpou@users.noreply.github.com>
Signed-off-by: malpou <malthe@grundtvigsvej.dk>

malpou committed:
fix e2e tests

Signed-off-by: Malthe Poulsen <30603252+malpou@users.noreply.github.com>
Signed-off-by: malpou <malthe@grundtvigsvej.dk>

malpou committed:
simplify e2e tests

Signed-off-by: Malthe Poulsen <30603252+malpou@users.noreply.github.com>
Signed-off-by: malpou <malthe@grundtvigsvej.dk>

malpou committed:
remove sleep

Signed-off-by: Malthe Poulsen <30603252+malpou@users.noreply.github.com>
Signed-off-by: malpou <malthe@grundtvigsvej.dk>

malpou committed:
refinement

Signed-off-by: Malthe Poulsen <30603252+malpou@users.noreply.github.com>
Signed-off-by: malpou <malthe@grundtvigsvej.dk>

malpou committed:
Update interceptor/handler/placeholder.go

Co-authored-by: Jorge Turrado Ferrero <Jorge_turrado@hotmail.es>
Signed-off-by: Malthe Poulsen <30603252+malpou@users.noreply.github.com>
Signed-off-by: malpou <malthe@grundtvigsvej.dk>

malpou committed:
Update tests/checks/placeholder_pages/placeholder_pages_test.go

Co-authored-by: Jorge Turrado Ferrero <Jorge_turrado@hotmail.es>
Signed-off-by: Malthe Poulsen <30603252+malpou@users.noreply.github.com>
Signed-off-by: malpou <malthe@grundtvigsvej.dk>

malpou committed:
linting issues

Signed-off-by: malpou <malthe@grundtvigsvej.dk>

JorTurFer commented on interceptor/handler/placeholder.go:
```diff
@@ -0,0 +1,308 @@
+			return nil, err
+		}
+
+		h.cacheMutex.Lock()
```
same as above

JorTurFer reviewed (commented):
awesome job! Only one small nit inline and it's okey to merge ❤️

Nico385412 commented:
Hello any news on this feature ? 🙏

wozniakjan reviewed (commented):
this looks really great, thank you for contributing!

I do have a couple of discussion points before it gets merged
* usage of `ConfigMaps` for storing content for placeholder pages - imho defaults would be better configured globally as a statically mounted `ConfigMap` or env variable on `Deployment`
* simplifying content processing - the default as a nicely formatted HTML page makes great sense but then we should allow arbitrary payloads with arbitrary headers, not all usage of this tool is aimed at web browsers, sometimes it's microservices talking between each other

malpou committed:
Update interceptor/handler/placeholder.go

Co-authored-by: Jan Wozniak <wozniak.jan@gmail.com>
Signed-off-by: Malthe Poulsen <30603252+malpou@users.noreply.github.com>
Signed-off-by: malpou <malthe@grundtvigsvej.dk>

malpou committed:
refactor: Address maintainer feedback on placeholder pages

- Move default templates to configuration via environment variables
- Add ConfigMap for default placeholder templates
- Support multiple content types (HTML, JSON, XML, plain text)
- Remove ConfigMap fields from HTTPScaledObject CRD
- Add error handling for endpoint cache failures
- Update tests to reflect all changes

This addresses all review comments:
- Defaults configured globally via mounted ConfigMap
- Arbitrary content types supported for microservices
- No RBAC changes needed (removed ConfigMap access)
- Proper error handling when cache fails

Signed-off-by: malpou <malthe@grundtvigsvej.dk>

malpou committed:
pr feedback adjustments [not tested yet]

Signed-off-by: malpou <malthe@grundtvigsvej.dk>

malpou commented:
@wozniakjan I've looked at the feedback and tried to adjust it accordingly.

Haven't had time to do the E2E tests, so put the PR in draft until it's ready for review.

malpou committed:
feat: make placeholder pages content-agnostic

Refactored placeholder pages to support any content type without
automatic modifications. Users can now provide HTML, JSON, XML, or
plain text content that will be served as-is during scale-from-zero.

Changes:
- Removed automatic script injection and default HTML template
- Added Content-Type header support in placeholderConfig
- Template variables (ServiceName, Namespace, RefreshInterval) work
  with any content format
- Added examples for JSON, XML, and plain text responses
- Fixed e2e tests and improved test helper reliability

Signed-off-by: malpou <malthe@grundtvigsvej.dk>

malpou committed:
fix: add defensive nil check for placeholderHandler

Reorder condition checks to verify placeholder configuration before
checking handler to prevent potential panics if placeholder is enabled
in spec but handler is nil.

Signed-off-by: malpou <malthe@grundtvigsvej.dk>

malpou committed:
refactor: improve placeholder handler code quality

Improve maintainability and readability of placeholder pages feature:

- Add constants for headers and messages to eliminate magic strings
- Remove self-explanatory comments that duplicate code intent
- Extract helper methods to reduce duplication and improve separation of concerns
- Rename getTemplate to resolveTemplate for clarity
- Simplify template caching with cleaner double-check locking pattern
- Extract placeholder serving logic into focused helper functions

No functional changes - pure refactoring for long-term maintainability.

Signed-off-by: malpou <malthe@grundtvigsvej.dk>

starlightromero commented:
Gentle bump @malpou . Looking forward to this feature

malpou committed:
test: add E2E tests for placeholder pages feature

Add 4 new test cases:
- testPlaceholderDisabled: backward compatibility test
- testTimestampTemplateVariable: validates {{.Timestamp}} substitution
- testRequestIDTemplateVariable: validates {{.RequestID}} from header
- testPlaceholderToRealServiceTransition: tests placeholder to service transition

Fix unused parameter lint warning in interceptor/main.go.

Signed-off-by: malpou <malthe@grundtvigsvej.dk>

malpou commented:
@starlightromero thx for the bump, good time to do it since i had no plans today 🙌🏼 

@JorTurFer / @wozniakjan can i get an CI execution to see if everything pass? I've gotten E2E running locally and from my perspective things look alright.

Would be good to reach a point where we want to merge this. I'm happy to allocate time to adjust the feedback you're having the coming month (hopefully we can reach a good point in a shorter time span).

malpou committed:
fix: resolve CI failures for placeholder pages PR

- Add trailing newlines to placeholder-templates.yaml and httpscaledobject.yaml
- Sort CHANGELOG entries alphabetically
- Update PlaceholderConfig Go doc comments to match CRD descriptions

Signed-off-by: Malthe Poulsen <malthe@grundtvigsvej.dk>
Signed-off-by: malpou <malthe@grundtvigsvej.dk>

linkvt commented:
Hi @malpou , I was finally working on this and created a new PR due to the many conflicts caused by our other changes, I'm working on it as part of the whole static responses topic.

I chose a different approach in a few areas and explained my reasoning in the PR description, in case you find some time to give it a look I would be happy to hear your thoughts, even if this PR is almost a year old and things might be a bit blurry.
Most of the decisions I made came from a defensive approach about how the code behaves in large clusters and from a YAGNI perspective, if you or another reader disagrees and has a valid argument things can of course be changed 🙂 .

linkvt commented:
Closing as the alternative placeholder response implementation #1626 was merged, see e.g. the docs: https://keda.sh/http-add-on/0.15/user-guide/configure-cold-start/#response-body-from-a-configmap

Please open an issue if you miss something!
